air/src/air/transition/frame.rs

```diff
@@ -77,3 +79,65 @@ impl<E: FieldElement> EvaluationFrame<E> {
         &mut self.next
     }
 }
+
+// LAGRANGE KERNEL EVALUATION FRAME
+// ================================================================================================
+
+/// The evaluation frame for the Lagrange kernel.
+///
+/// The Lagrange kernel's evaluation frame is different from [`EvaluationFrame`]. Specifically,
+/// - it only contains evaluations from the Lagrange kernel column
+///   (compared to all columns in the case of [`EvaluationFrame`])
+/// - The column is evaluated at points `z`, `gz`, `g^2 z`, ..., `g^(2^(v-1)) z`,
+///   where `z` is an arbitrary point, and `g` is the trace domain generator
+#[derive(Debug, Clone)]
+pub struct LagrangeKernelEvaluationFrame<E: FieldElement> {
+    frame: Vec<E>,
+}
+
+impl<E: FieldElement> LagrangeKernelEvaluationFrame<E> {
+    // CONSTRUCTORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Constructs a Lagrange kernel evaluation frame from the raw column polynomial evaluations.
+    pub fn new(frame: Vec<E>) -> Self {
+        Self { frame }
+    }
+
+    /// Constructs the frame from the Lagrange kernel trace column polynomial coefficients for an
+    /// evaluation point.
+    pub fn from_lagrange_kernel_column_poly(lagrange_kernel_col_poly: &[E], z: E) -> Self {
+        let log_trace_len = lagrange_kernel_col_poly.len().ilog2();
+        let g = E::from(E::BaseField::get_root_of_unity(log_trace_len));
+
+        let mut frame = Vec::with_capacity(log_trace_len as usize + 1);
+
+        // push c(z)
+        frame.push(polynom::eval(lagrange_kernel_col_poly, z));
+
+        // push `c(gz)`, `c(z * g^2)`, `c(z * g^4)`, ..., `c(z * g^(2^(v-1)))`
+        for i in 0..log_trace_len {
+            let x = g.exp_vartime(2_u32.pow(i).into()) * z;
```

Comment:
Actually we're increasing the power nonlinearly. The powers are: 1, 2, 4, 8, 16, ... . We could try to optimize this a bit, but we're only doing this `log(trace_len)` times, i.e. always a relatively few number of times. I think the added complexity wouldn't be worth the marginal performance improvement?

Reply:
Since this is used only for the OOD frame, using `exp_vartime()` is probably fine - but also, I don't think this would add much complexity.

If the powers are $g$, $g^2$, $g^4$, $g^8$ etc. wouldn't this be just:
```Rust
let mut g_exp = g;
for i in 0..log_trace_len {
    let x = g_exp * z;
    let lagrange_poly_at_x = polynom::eval(lagrange_kernel_col_poly, x);

    frame.push(lagrange_poly_at_x);
    g_exp *= g_exp;
}
```

Reply:
bleh, you're right. I changed it to use this code snippet.